rust_src/src/windows.rs

```diff
@@ -179,3 +184,17 @@ pub fn minibuffer_selected_window() -> LispObject {
         LispObject::constant_nil()
     }
 }
+
+/// Return the frame that window WINDOW is on.
+/// WINDOW is optional and defaults to the selected window. If provided it must
+/// be a valid window.
+#[lisp_fn(min = "0")]
+pub fn window_frame(window: LispObject) -> LispObject {
+    let win = if window.is_nil() {
+        selected_window()
+    } else {
+        window
+    };
+
+    win.as_window_or_error().frame()
+}
```

Comment:
It looks like the c code makes use of `CHECK_VALID_WINDOW` which throws a `Qwindow_valid_p` error. We don't have this implemented yet, but I wonder if we already have ported functions with `decode_valid_window` incorrectly, hmmm.

Reply:
I have a PR almost ready :-)

Reply:
The window-buffer routines also should have used `CHECK_VALID`

Reply:
I am adding `window_valid_or_selected` and `window_or_selected` to windows.rs.

Reply:
Bike shed while I am away. I will commit this afternoon.